Framework/Algorithms/src/AppendSpectra.cpp

```diff
@@ -131,10 +133,46 @@ void AppendSpectra::fixSpectrumNumbers(API::MatrixWorkspace_const_sptr ws1,
   if (ws2min > ws1max)
     return;
 
-  // change the axis by adding the maximum existing spectrum number to the
-  // current value
-  for (size_t i = 0; i < output->getNumberHistograms(); i++)
-    output->getSpectrum(i).setSpectrumNo(specnum_t(i));
+  const int yAxisNum = 1;
+  const auto yAxisWS1 = ws1->getAxis(yAxisNum);
+  const auto yAxisWS2 = ws2->getAxis(yAxisNum);
+  auto outputYAxis = output->getAxis(yAxisNum);
+  auto ws1len = ws1->getNumberHistograms();
+  // if spectra numbers
+
+  const bool isTextAxis = (yAxisWS1->isText() && yAxisWS2->isText());
+  const bool isNumericAxis = (yAxisWS1->isNumeric() && yAxisWS2->isNumeric());
+
+  if (isTextAxis) {
+    // have to cast up to TextAxis, setLabel is TextAxis specific method
+    auto outputTextAxis = dynamic_cast<TextAxis *>(outputYAxis);
+
+    for (size_t i = 0; i < output->getNumberHistograms(); i++) {
+      // check if we're outside the spectra of the first workspace
+      std::string inputLabel =
+          i < ws1len ? yAxisWS1->label(i) : yAxisWS2->label(i - ws1len);
+
+      outputTextAxis->setLabel(i, (inputLabel.size() > 0) ? inputLabel : "");
+      output->getSpectrum(i).setSpectrumNo(specnum_t(i));
+    }
+  } else if (isNumericAxis) {
+
+    for (size_t i = 0; i < output->getNumberHistograms(); i++) {
+      // check if we're outside the spectra of the first workspace
+      double inputVal =
+          i < ws1len ? yAxisWS1->getValue(i) : yAxisWS2->getValue(i - ws1len);
+      outputYAxis->setValue(i, inputVal);
+      output->getSpectrum(i).setSpectrumNo(specnum_t(i));
+    }
+
+  } else {
+    // change the axis by adding the maximum existing spectrum number to the
+    // current value
+    // default to old behaviour if none of our special cases are true
+    for (size_t i = 0; i < output->getNumberHistograms(); i++) {
+      output->getSpectrum(i).setSpectrumNo(specnum_t(i));
+    }
```

Comment:
If I understood the above code right, you could also consider writing this if-else if-else thing the other way around:
```C++
  for (size_t i = 0; i < output->getNumberHistograms(); ++i) {
    if (isTextAxis) {
      const std::string inputLabel = i < ws1len ? yAxisWS1->label(i) : yAxisWS2->label(i - ws1len);
      outputTextAxis->setLabel(i, (inputLabel.size() > 0) ? inputLabel : "");
    }
    else if (isNumericAxis) {
      const double inputVal = i < ws1len ? yAxisWS1->getValue(i) : yAxisWS2->getValue(i - ws1len);
      outputYAxis->setValue(i, inputVal);
    }
    output->getSpectrum(i).setSpectrumNo(specnum_t(i));
  }
```
Less code duplication, no?